insights/combiners/modprobe.py

```diff
@@ -0,0 +1,89 @@
+from insights.parsers.modprobe import ModProbe
+from .. import LegacyItemAccess
+
+
+@combiner(requires=[ModProbe])
+class AllModProbe(LegacyItemAccess):
+    """
+    Combiner for accessing all the modprobe configuration files in one
+    structure.
+
+    It's important for our reporting and information purposes to know not
+    only what the configuration was but where it was defined.  Therefore, the
+    format of the data in this combiner is slightly different compared to the
+    ModProbe parser.  Here, each 'value' is actually a 2-tuple, with the
+    actual data first and the file name from whence the value came second.
+    This does mean that you need to pull the value out of each item - e.g.
+    using a list comprehension - but it means that every item is associated
+    with the file it was defined in.
+
+    In line with the ModProbe configuration parser, the actual value is
+    usually a list of the space-separated parts on the line, and the
+    definitions for each module are similarly kept in a list, which makes
+
+    Thanks to the LegacyItemAccess class, this can also be treated as a
+    dictionary for look-ups of data in the `data` attribute.
+
+    Attributes:
+        data (dict): The combined data structures, with each item as a
+            2-tuple, as described above.
+        bad_lines(list): The list of unparseable lines from all files, with
+            each line as a 2-tuple as described above.
+
+    Sample data files::
+
+        /etc/modprobe.conf:
+            # watchdog drivers
+            blacklist i8xx_tco
+
+            # Don't install the Firewire ethernet driver
+            install eth1394 /bin/true
+
+        /etc/modprobe.conf.d/no_ipv6.conf:
+            options ipv6 disable=1
+            install ipv6 /bin/true
+
+    Examples:
+        >>> all_modprobe = shared[AllModProbe]
+        >>> all_modprobe['alias']
+        []
+        >>> all_modprobe['blacklist']
+        {'i8xx_tco': (True, '/etc/modprobe.conf')}
+        >>> all_modprobe['install']
+        {'eth1394': (['/bin/true'], '/etc/modprobe.conf'),
+         'ipv6': (['/bin/true'], '/etc/modprobe.conf.d/no_ipv6.conf')}
+    """
+    def __init__(self, local, shared):
+        self.data = {}
+        self.bad_lines = []
+        for mod in shared[ModProbe]:
+            filename = mod.file_path  # relative path inside archive
+            # Copy data section
+            for section, sectdict in mod.data.iteritems():
+                if section not in self.data:
+                    self.data[section] = {}
+                for name, value in sectdict.iteritems():
+                    if name in self.data[section]:
+                        # append to this module's value - should only
+                        # happen for aliases.
+                        self.data[section][name][0].append(value)
+                    else:
+                        # create new tuple
+                        self.data[section][name] = (value, filename)
```

Comment:
@PaulWay have you considered using a `namedtuple` type to make it a bit clearer regarding the data contents and how to access them?

Reply:
OK, I've used a named tuple, but the documentation is playing up on me.  I have:
```python
ModProbeValue = namedtuple("ModProbeValue", ['value', 'source'])
"""
A value from a ModProbe source

Attributes:
    value (object): The original value stored in the source file, or a list
        of values if this is an alias.
    source (str): The source file name of the associated value.
"""
```

But it lists the value and source attributes twice.  Is there a better way?

Reply:
When you have an `__init__` method in your class you should document attributes there with docstrings following the attribute initialization instead of in the class docstring.  See these examples in this reference http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html:
```
    #: list of str: Doc comment *before* attribute, with type specified
    self.attr4 = ['attr4']

     self.attr5 = None
     """str: Docstring *after* attribute, with type specified."""
```

Reply:
Sorry, misunderstood what you were asking.  I think you should just not use the Attributes tag.